digest/src/newtype.rs

```diff
@@ -0,0 +1,528 @@
+                        $inner,
+                    > as $crate::crypto_common::hazmat::SerializableState>::SerializedStateSize;
+
+                    #[inline]
+                    fn serialize(&self) -> $crate::crypto_common::hazmat::SerializedState<Self> {
+                        self.0.serialize()
+                    }
+
+                    #[inline]
+                    fn deserialize(
+                        serialized_state: &$crate::crypto_common::hazmat::SerializedState<Self>,
+                    ) -> Result<Self, $crate::crypto_common::hazmat::DeserializeStateError> {
+                        Ok(Self($crate::newtype::Wrapper::deserialize(
+                            serialized_state,
+                        )?))
+                    }
+                }
+            };
+        }
+    }
+
+    #[macro_export]
+    macro_rules! newtype_common {
+        ($name: ident, $inner: ty) => {
+            $crate::newtype_common_customized_init!($name, $inner);
+            $crate::newtype_common_extendable_output!($name, $inner);
+            $crate::newtype_common_extendable_output_reset!($name, $inner);
+            $crate::newtype_common_fixed_output!($name, $inner);
+            $crate::newtype_common_fixed_output_reset!($name, $inner);
+            $crate::newtype_common_hash_marker!($name, $inner);
+            $crate::newtype_common_inner_user!($name, $inner);
+            $crate::newtype_common_output_size_user!($name, $inner);
+            $crate::newtype_common_reset!($name, $inner);
+            $crate::newtype_common_update!($name, $inner);
+            $crate::newtype_common_serializable_state!($name, $inner);
+        };
+    }
+}
+
+#[cfg(feature = "mac")]
+mod mac {
+    use super::Wrapper;
+    use crate::{InnerInit, MacMarker};
+    use core::marker::PhantomData;
+
+    impl<T> InnerInit for Wrapper<'_, T>
+    where
+        T: InnerInit,
+    {
+        fn inner_init(inner: Self::Inner) -> Self {
+            Wrapper {
+                inner: T::inner_init(inner),
+                _marker: PhantomData,
+            }
+        }
+    }
+
+    impl<T> MacMarker for Wrapper<'_, T> where T: MacMarker {}
+
+    #[macro_export]
+    macro_rules! newtype_mac {
+        ($name: ident, $inner: ty) => {
+            impl $crate::InnerInit for $name
+            where
+                for<'a> $crate::newtype::Wrapper<'a, $inner>: $crate::InnerInit,
+            {
+                #[inline]
+                fn inner_init(inner: Self::Inner) -> Self {
+                    Self($crate::newtype::Wrapper::<'static, $inner>::inner_init(
+                        inner,
+                    ))
+                }
+            }
+        };
+    }
+}
+
+#[cfg(not(feature = "mac"))]
+#[macro_export]
+macro_rules! newtype_mac {
+    ($name: ident, $inner: ty) => {};
+}
+
+#[cfg(feature = "oid")]
+mod oid {
+    use super::Wrapper;
+    use crate::const_oid::{AssociatedOid, ObjectIdentifier};
+
+    impl<T> AssociatedOid for Wrapper<'_, T>
```

Comment:
I think it'd be nice to get rid of all of the current complexity we have around this trait (which is really just a hack around the fact that the digest crates couldn't define their own types) and just simply impl `AssociatedOid` for the newtype.

Reply:
Ha good point, I just used the same pattern mechanically.